backpack/core/derivatives/crossentropyloss.py

```diff
@@ -145,3 +178,120 @@ def _check_2nd_order_parameters(self, module):
+
+        Undoes treating additional axes like batch axis and assumes an number of
+        additional free axes (``v``) were added, i.e. un-groups
+        ``v (n d1 d2) c -> v n c d1 d2``.
+
+        Args:
+            tensor: the tensor to rearrange
+            str_d_dims: a string representation of the additional dimensions
+            d_info: a dictionary encoding the size of the additional dimensions
+            free_axis: Number of free leading axes. Default: ``1``.
+
+        Returns:
+            the rearranged tensor
+
+        Raises:
+            NotImplementedError: If ``free_axis != 1``.
+        """
+        if free_axis != 1:
+            raise NotImplementedError(f"Only supports free_axis=1. Got {free_axis}.")
+
+        return rearrange(
+            tensor, f"v (n {str_d_dims}) c -> v n c {str_d_dims}", **d_info
+        )
+
+    @staticmethod
+    def _expand_sqrt_h(
+        sqrt_h: Tensor, module: CrossEntropyLoss, subsampling: Union[None, List[int]]
+    ) -> Tensor:
+        """Expands the square root hessian if CrossEntropyLoss has additional axes.
+
+        In the case of e.g. two additional axes, the input is [N,C,A,B].
+        In CrossEntropyLoss the additional axes are treated independently.
+        Therefore the intermediate result has shape [C,N,C,A,B].
+        In subsequent calculations the additional axes are not independent anymore.
+        The required shape for sqrt_h_full is then [C*A*B,N,C,A,B].
+        Due to the independence, sqrt_h lives on the diagonal of sqrt_h_full.
+
+        Args:
+            sqrt_h: intermediate result, shape [C,N,C,A,B]
+            module: module, to extract input shape
+            subsampling: subsampling to determine length of batch axis
+
+        Returns:
+            sqrt_h_full, shape [C*A*B,N,C,A,B], sqrt_h on diagonal.
+        """
+        n_standard_axes = 2
+        n_additional_axes = module.input0.dim() - n_standard_axes
+
+        if n_additional_axes > 0:
+            shape_input = module.input0.shape
+            N_samples = shape_input[0] if subsampling is None else len(subsampling)
+            sqrt_H_full = zeros(*shape_input[1:], N_samples, *shape_input[1:])
+            ranges_additional_axes: List[Iterable, ...] = [
+                range(shape_input[n_standard_axes + i])
+                for i in range(n_additional_axes)
+            ]
+            index_additional_axes: Tuple[int]
+            for index_additional_axes in itertools_product(*ranges_additional_axes):
+                selection_sqrt_H = (slice(None),) * 3 + index_additional_axes
+                selection_target = (
+                    (slice(None),)
+                    + index_additional_axes
+                    + (slice(None),) * 2
+                    + index_additional_axes
+                )
+                sqrt_H_full[selection_target] = sqrt_h[selection_sqrt_H]
+            return sqrt_H_full.flatten(0, n_additional_axes)
+        return sqrt_h
```

Comment:
I think that, based on Tuesday's write-up, there should be a way to implement this cleaner, using [`torch.diag_embed`](https://pytorch.org/docs/stable/generated/torch.diag_embed.html) and no `for` loops.

Reply:
I also feel like there should be a way. But after thinking for a while, I don't think there is a way.
`torch.diag_embed` does a 2D embedding. Since this is a multidimensional embedding, it would have to be a concatenation of 2D-diagonal embeddings. Do you think this is faster than the for-loop? (My guess is, that it does just the same)

Reply:
The documentation says that (d>1)-dimensional inputs can also be embedded. My guess is that if we can use `diag_embed` and some `reshape`s, this should be faster than a for loop.

Reply:
Indeed, this works.
Using flatten - diag_embed - reshape.